src/constraints/constraints.jl

```diff
@@ -0,0 +1,247 @@
+    idx_dist_pairs = spzeros(n_atoms, n_atoms) * unit(constraints[1].dist)
+
+    # Store constraints as directed edges, direction is arbitrary but necessary
+    for constraint in constraints
+        edge_added = add_edge!(constraint_graph, constraint.atom_idxs[1], constraint.atom_idxs[2])
+        if edge_added
+            idx_dist_pairs[constraint.atom_idxs[1],constraint.atom_idxs[2]] = constraint.dist
+            idx_dist_pairs[constraint.atom_idxs[2],constraint.atom_idxs[1]] = constraint.dist
+        else
+            @warn "Duplicated constraint in System. It will be ignored."
+        end
+    end
+
+    # Get groups of constraints that are connected to eachother 
+    cc = connected_components(constraint_graph) #& this can return non-connected things sometimes??
+    # Initialze empty vector of clusters
+    clusters = Vector{ConstraintCluster}(undef, 0)
+
+    #Loop through connected regions and convert to clusters
+    for (cluster_idx, atom_idxs) in enumerate(cc)
+        # Loop atoms in connected region to build cluster
+        if length(atom_idxs) > 1 #connected_components gives unconnected verticies as well
+            connected_constraints = Vector{DistanceConstraint}(undef, 0)
+            for ai in atom_idxs
+                neigh_idxs = neighbors(constraint_graph, ai)
+                for neigh_idx in neigh_idxs
+                    push!(connected_constraints,
+                        DistanceConstraint(SVector(ai, neigh_idx), idx_dist_pairs[ai,neigh_idx]))
+                end
+            end
+            connected_constraints = convert(SVector{length(connected_constraints)}, connected_constraints)
+            push!(clusters, ConstraintCluster(connected_constraints))
+        end
+    end
+
+    return clusters
+end
+
+
+##### High Level Constraint Functions ######
+function save_ca_positions!(sys::System, c)
+    if length(sys.constraints) > 0
+        sys.hidden_storage.ca_coord_storage .= c
+    end
+end
+
+function apply_position_constraints!(sys::System, coord_storage;
+     n_threads::Integer=Threads.nthreads())
+
+   for ca in sys.constraints
+       position_constraints!(sys, ca, coord_storage, n_threads = n_threads)
+   end
+
+   return sys
+
+end
+
+function apply_position_constraints!(sys::System, coord_storage,
+     vel_storage, dt; n_threads::Integer=Threads.nthreads())
+
+     if length(sys.constraints) > 0
+        vel_storage .= -sys.coords ./ dt
+
+        for ca in sys.constraints
+            position_constraints!(sys, ca, coord_storage, n_threads = n_threads)
+        end
+
+        vel_storage .+= sys.coords ./ dt
+
+        sys.velocities .+= vel_storage
+    end
+
+   return sys
+end
+
+
+function apply_velocity_constraints!(sys::System; n_threads::Integer=Threads.nthreads())
+    
+    for ca in sys.constraints
+        velocity_constraints!(sys, ca, n_threads = n_threads)
+    end
+
+    return sys
+end
+
+
+function check_position_constraints(sys::System, ca::ConstraintAlgorithm)
```

Comment:
max_err is just the error I calculate, the actual tolerance is already in the constraint algorithm (`ca.tolerance`). Since the constraints are iterative you always should be able to reach that tol.

Reply:
:+1: